### PR TITLE
[FLINK-23801][connector/kafka] Report numBytesIn and pendingRecords in KafkaSource

### PR DESCRIPTION
**Pull request by PatrickRen**
## What is the purpose of the change

This pull requests add metric "numBytesIn" and "pendingRecords" for KafkaSource

## Brief change log
- Register "numBytesIn" and "pendingRecords" for KafkaSource

## Verifying this change
This change added tests and can be verified as follows:
- Create ```KafkaPartitionSplitReader``` and assign splits
- Fetch records and check if metrics work s expected

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:437]: I wonder if this is the correct metric to track. AFAIK this metric describes the following 
`The maximum lag in terms of number of records for any partition in this window` so every subtask will show the same lag value. 

I know there is a similar metric like `records-lag` which is defined per partition. One could compute the sum of all the currently assigned partitions to get a proper lag calculation.

Unfortunately, the last time I looked into these metrics I remember vaguely that the _per-partition_ metrics are only registered after one record was consumed.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:178]: I have not much experience with how Kafka records are structured but did you consider also counting the metadata fields?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/CommonTestUtils.java
Comment (fapaul) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/CommonTestUtils.java:206]:
```suggestion
        boolean conditionResult = condition.get();
```
Is it necessary to do one loop here?

Comment (fapaul) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/CommonTestUtils.java:206]: This also seems like a bugfix commit :)

**Review by fapaul — COMMENTED** [review]
Thanks for the contribution. I quickly double-checked with FLIP-33 and AFAICT you have covered all the metrics we can implement. 
I left some questions regarding how we derive and compute the metrics.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:178]: Maybe it is better to directly forward the Kafka metric https://github.com/apache/flink/commit/a1441c262f35407c4029f6ddf8daa18cd859f822#diff-026b7f554eb085f1d2dde933dcafa4bf39abd995dcb071126a545c9d9de109fbR41

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:437]: Thanks for the careful review! This is really annoying... I'm not sure if that worth it to use additional data structures and maybe affect performance to implement just one metric. Is there any suggestions for ```pendingRecords```? Otherwise I prefer to skip it for now 😢

Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:437]: I did another round of digging in Kafka and my initial assumption was not fully correct the `records-lag-max` metric is the maximum lag of all partitions assigned to the current consumer.
This gives us the possibility to simply multiply the current partition count by the lag value and we have a rough estimation.
WDYT?

Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java]: I think multiplying with #partition is a bit too inaccurate. If messages among subscribed partitions are imbalanced, the value of ```pendingRecords``` will be quite misleading to users.

Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java]: It can definitely be misleading when multiplying it in the case you've mentioned but on the other hand in the current situation, it is misleading if all assigned partitions lag equally behind. The `pendingRecords` would only account for one partition although the correct lag is up to partitions * `records-max-lag`. 

Personally, I can also think of not implementing this metric now and separate it from this PR. 
WDYT @AHeise?

**Comment by PatrickRen** [comment]
Thanks for the review @fapaul ! I made another push with some changes based on your comments. About ```pendingRecords ``` I'd like to postpone it until we have a better solution to implement this. Could you please have another review? Thanks!

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:436]: Nit: You can directly return the `Optional<Supplier<Long>>` to replace the null return.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:437]: Nit: extract constants for the metric name and the name of the metric group

**Review by fapaul — APPROVED** [review]
Thanks for quickly addressing the feedback I left some minor comments but overall the PR looks ready to go :shipit:

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:436]: The compiler complains a warn if I use ```Optional``` as a class field 🤔 I'd like to keep it for now (forgive me being neat-freak lol)

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java]: Honestly, you are thinking too hard. Simply expose `records-max-lag` as is. That's the metric that kafka exposes for it's consumer/fetcher, so that's exactly what Kafka users expect to see as a lag.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:90]: I don't immediately see the value of using the `Supplier` indirection over using `@Nullable Metric bytesConsumedMetric`. Having a nullable supplier is kind of an anti-pattern as we are mixing functional-style with imperative constructs (null).

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:453]: I'd extract that to `Optional<Metric> findKafkaMetic(String group, String name)`, then you can reuse it for the lag metric.

**Review by AHeise — COMMENTED** [review]
Imho we could add the lag as is. I have also added two suggestions.

Btw we don't need to rush things here. If this is the last thing that we want to add for Kafka, I'm sure we can apply for an exceptional merge after feature freeze as this will definitively not break anything.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:136]: Can you rebase and check if this constructor is still public?

Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:136]: I have done https://github.com/fapaul/flink/commit/01de3e6db4c68d316d02bda35279e163673aa2bc to circumvent the problem I think you can do similarly for the source metric group.

Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java]: Thanks @fapaul for the reminder! I made another detour in 744fcd3

**Comment by PatrickRen** [comment]
Thanks @fapaul @AHeise for the review~ I re-implemented pendingRecords and numBytesIn in the latest push. I implement these metrics in ```KafkaSourceReaderMetrics``` to make ```KafkaPartitionSplitReader``` clearer. Could you please have a review again? Thanks!

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:208]: Why is this done lazily?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:216]: If I understand this code correctly we sum the lag of all assigned partitions. What is the difference between doing so and just using `records-lag-max`?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:142]: I think it is better to introduce a new method `InternalSourceReaderMetricGroup#mock(MetricGroup metricGroup, OperatorIOMetricGroup opMetricGroup)` like I did here https://github.com/fapaul/flink/commit/01de3e6db4c68d316d02bda35279e163673aa2bc and not use the task metric.

**Review by fapaul — CHANGES_REQUESTED** [review]
I think we can unify the metric handling across the source and sink. Unfortunately, I still did not fully understand the lag calculation maybe you can shed some light on why we need to do this way.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:216]: I now understand the calculation :)

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaConsumerMetricUtils.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaConsumerMetricUtils.java:29]: That's a good one. I'll pick your implementation for now and rebase after your PR is merged.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:208]: The value of ```pendingRecords``` can only be known after the first ```consumer.poll()```. If this metric is registered right on start, we need to assign an initial value like 0 or -1 before the first poll. 0 might be confusing since it also means "no lag", and -1 is kinda weird.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:142]: What about using Mockito? 

```java
// Mock an operator metric group for intercepting numBytesIn
final OperatorMetricGroup operatorMetricGroup =
        Mockito.mock(OperatorMetricGroup.class, Mockito.RETURNS_DEEP_STUBS);
Counter numBytesInCounter = new SimpleCounter();
Mockito.when(operatorMetricGroup.getIOMetricGroup().getNumBytesInCounter())
        .thenReturn(numBytesInCounter);
SourceReaderMetricGroup sourceReaderMetricGroup =
        InternalSourceReaderMetricGroup.wrap(operatorMetricGroup);
```

This can avoid implementing another ```NumBytesInOperatorIOMetricGroup``` with a bunch of ```return null``` stuff.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (AHeise) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:249]: Can we please avoid such large lambdas by pulling out method lambdas?

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (AHeise) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:268]: I think we also need sum up the metrics per split, right? Currently, if we have two splits and consumers, they would overwrite each other.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java
Comment (AHeise) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReaderTest.java:144]: Mockito is banned for new stuff since we got burnt several times in the past.
#16875 now managed to just use the existing stuff, so please check if you can use that.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/metrics/testutils/MetricListener.java
Comment (AHeise) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/metrics/testutils/MetricListener.java:78]: Use Optional?

**Review by AHeise — CHANGES_REQUESTED** [review]
I have some concerns about the bytes-in metric and [Mockito is banned](https://flink.apache.org/contributing/code-style-and-quality-common.html#design-for-testability) for new tests afaik.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java
Comment (PatrickRen) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java:268]: I think ```bytes-consumed-total``` without topic in the tag already sums them up. Also there should be only one consumer instance running in one subtask, so there won't be overwrite I think

Comment (AHeise) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetrics.java]: Sorry I was confused. I thought there was a `SplitReader` per split but that's obviously not the case.

**Review by AHeise — APPROVED** [review]
LGTM.

**Comment by preston-m-price** [comment]
@PatrickRen can you take a look at whether this PR introduced this issue: https://issues.apache.org/jira/browse/FLINK-24497?

**Pull request merged**
